pkgs/cli_config/lib/src/config.dart

```diff
@@ -0,0 +1,387 @@
+import 'file_provider.dart';
+import 'provider.dart';
+
+/// A hierarchical configuration object.
+///
+/// Configuration can be provided as commandline arguments, environment
+/// variables and configuration files. This configuration object makes
+/// these accessible with a uniform API.
+///
+/// Configuration can be provided in three ways:
+/// 1. commandline argument defines as `-Dsome_key=some_value`,
+/// 2. environment variables as `SOME_KEY=some_value`, and
+/// 3. config files as JSON or YAML as `{'some_key': 'some_value'}`.
+///
+/// The default lookup behavior is that commandline argument defines take
+/// precedence over environment variables, which take precedence over the
+/// configuration file.
+///
+/// The config is hierarchical in nature, using `.` as the hierarchy separator
+/// for lookup and commandline defines. The hierarchy should be materialized in
+/// the JSON or YAML configuration file. For environment variables `__` is used
+/// as hierarchy separator.
+///
+/// Hierarchical configuration can be provided in three ways:
+/// 1. commandline argument defines as `-Dsome_key.some_nested_key=some_value`,
+/// 2. environment variables as `SOME_KEY__SOME_NESTED_KEY=some_value`, and
+/// 3. config files as JSON or YAML as
+///    ```yaml
+///    some_key:
+///      some_nested_key:
+///        some_value
+///    ```
+///
+/// The config is opinionated on the format of the keys.
+/// * Command-line argument keys should be lower-cased alphanumeric
+///   characters or underscores, with `.` for hierarchy.
+/// * Environment variables keys should be upper-cased alphanumeric
+///    characters or underscores, with `__` for hierarchy.
+/// * Config files keys should be lower-cased alphanumeric
+///   characters or underscores.
+///
+/// In the API they are made available lower-cased and with underscores.
+class Config {
+  final CliProvider _cliProvider;
+  final EnvironmentProvider _environmentProvider;
+  final FileProvider _fileProvider;
+
+  /// Config providers, ordered by precedence.
+  late final _providers = [_cliProvider, _environmentProvider, _fileProvider];
+
+  Config._(
+    this._cliProvider,
+    this._environmentProvider,
+    this._fileProvider,
+  );
+
+  /// Constructs a config by parsing the three sources.
+  ///
+  /// If provided, [cliDefines] must be a list of '<key>=<value>'.
+  ///
+  /// If provided, [environment] must be a map containing environment varibales.
+  ///
+  /// At most one of [fileContents] or [fileParsed] must be provided.
+  /// If provided, [fileContents] must be valid JSON or YAML.
+  /// If provided, [fileParsed] must be valid parsed YSON or YAML (maps, lists,
+  /// strings, integers, and booleans).
+  ///
+  /// If provided [fileSourceUri], is used to provide better error messages on
+  /// parsing the configuration file.
+  factory Config({
+    List<String> cliDefines = const [],
+    Map<String, String> environment = const {},
+    Map<String, dynamic>? fileParsed,
```

Comment:
I would make this more opinionated about parsed or not contents, it is an awkward api currently that can't statically enforce good usage.

If you really want to support both I would consider a wrapper class, with two constructors, that does the parsing in the constructor that takes raw contents. Or possibly multiple constructors on this class that are more specialized.